File: langs/typescript/src/composites/scoredFilteredCompositeVM.ts
/**
 * ScoredFilteredCompositeVM — score-ranked visible projection.
 */
import { ComponentVMBase } from "../components/componentVMBase.js";
import type { CompositeVMBase } from "./compositeVMBase.js";
import {
  FilteredCompositeVM,
  type FilteredCompositeOptions,
} from "./filteredCompositeVM.js";

export interface ScoredFilteredCompositeOptions<VM extends ComponentVMBase>
  extends Omit<FilteredCompositeOptions<VM>, "predicate"> {
  readonly scorer: (vm: VM) => number | null;
}

export class ScoredFilteredCompositeVM<VM extends ComponentVMBase> extends FilteredCompositeVM<VM> {
  readonly #scorer: (vm: VM) => number | null;

  constructor(
    source: CompositeVMBase<VM>,
    options: ScoredFilteredCompositeOptions<VM>,
  ) {
    const scorer = options.scorer;
    super(source, {
      predicate: (vm) => scorer(vm) !== null,
      deferInitialRecompute: true,
      ...(options.cursorPolicy === undefined ? {} : { cursorPolicy: options.cursorPolicy }),
    });
    this.#scorer = scorer;
    this._recompute();
  }

  protected override orderedVisible(): VM[] {
    return [...this.source]
      .map((vm, index) => ({ vm, index, score: this.#scorer(vm) }))
      .filter((entry): entry is { vm: VM; index: number; score: number } => entry.score !== null)
      .sort((a, b) => b.score - a.score || a.index - b.index)
      .map((entry) => entry.vm);
  }

  refreshScores(): void {
    this._recompute();
  }
}

File: langs/swift/Sources/VMx/Composites/ScoredFilteredCompositeVM.swift
//
// ScoredFilteredCompositeVM.swift — score-ranked visible projection.
//

public final class ScoredFilteredCompositeVM<Child: ComponentVMBase>: FilteredCompositeVM<Child> {
    private let scorer: (Child) -> Double?

    public init(
        _ source: CompositeVM<Child>,
        cursorPolicy: FilteredCursorPolicy = .snapToFirst,
        scorer: @escaping (Child) -> Double?
    ) {
        self.scorer = scorer
        super.init(
            source,
            cursorPolicy: cursorPolicy,
            predicate: { scorer($0) != nil },
            deferInitialRecompute: true
        )
        refreshScores()
    }

    public override func orderedVisible() -> [Child] {
        (0..<source.count)
            .map { (index: $0, vm: source.at($0), score: scorer(source.at($0))) }
            .filter { $0.score != nil }
            .sorted {
                if $0.score! == $1.score! { return $0.index < $1.index }
                return $0.score! > $1.score!
            }
            .map(\.vm)
    }

    public func refreshScores() {
        recompute()
    }
}
